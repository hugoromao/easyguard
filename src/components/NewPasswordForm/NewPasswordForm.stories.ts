import type { Meta, StoryObj } from "@storybook/react";

import NewPasswordForm from ".";

const meta: Meta<typeof NewPasswordForm> = {
  component: NewPasswordForm,
};

export default meta;

type Story = StoryObj<typeof NewPasswordForm>;

export const Primary: Story = {};
