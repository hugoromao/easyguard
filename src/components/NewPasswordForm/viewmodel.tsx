import { useState } from "react";
import { enqueueSnackbar } from "notistack";

import { validateUserNumbers, validateUserWords } from "../../utils/passwd";

import { Input } from "./model";

export const useNewPasswordFormViewModel = (
  // eslint-disable-next-line no-unused-vars
  onFinish: (words: Input[], numbers: Input[]) => void
) => {
  const [step, setStep] = useState(1);
  const [words, setWords] = useState<Input[]>([
    { value: "", error: undefined },
    { value: "", error: undefined },
    { value: "", error: undefined },
    { value: "", error: undefined },
  ]);
  const [numbers, setNumbers] = useState<Input[]>([
    { value: "", error: undefined },
    { value: "", error: undefined },
  ]);
  const [loading, setLoading] = useState(false);

  const nextStep = () => setStep((s) => s + 1);
  const prevStep = () => setStep((s) => s - 1);

  const clearErrors = () => {
    setWords((words) => words.map((w) => ({ ...w, error: undefined })));
    setNumbers((numbers) => numbers.map((w) => ({ ...w, error: undefined })));
  };

  const validateWords = () => {
    let hasErrors = false;
    words.forEach((word, index) => {
      if (word.value === "") {
        hasErrors = true;
        setWords((w) => {
          const temp = w;
          temp[index].error = "Este campo é obrigatório.";
          return [...temp];
        });
      }
    });

    if (words.findIndex((w) => w.value === "") !== -1) return;

    const errors = validateUserWords(words.map((w) => w.value));
    errors.forEach((err) => {
      hasErrors = true;
      if (err.index !== undefined) {
        setWords((w) => {
          const temp = w;
          temp[err.index!].error = err.message;
          return [...temp];
        });
      } else {
        enqueueSnackbar(err.message, { variant: "error" });
      }
    });

    if (!hasErrors) nextStep();
  };

  const validateNumbers = () => {
    let hasErrors = false;
    numbers.forEach((number, index) => {
      if (number.value === "") {
        hasErrors = true;
        setNumbers((w) => {
          const temp = w;
          temp[index].error = "Este campo é obrigatório.";
          return [...temp];
        });
      }

      if (isNaN(Number(number.value))) {
        hasErrors = true;
        setNumbers((w) => {
          const temp = w;
          temp[index].error = "Número inválido.";
          return [...temp];
        });
      }
    });

    if (
      numbers.findIndex((n) => n.value === "") !== -1 ||
      numbers.findIndex((n) => isNaN(Number(n.value))) !== -1
    )
      return;

    const errors = validateUserNumbers(numbers.map((n) => Number(n.value)));
    errors.forEach((err) => {
      hasErrors = true;
      if (err.index !== undefined) {
        setNumbers((w) => {
          const temp = w;
          temp[err.index!].error = err.message;
          return [...temp];
        });
      } else {
        enqueueSnackbar(err.message, { variant: "error" });
      }
    });

    if (!hasErrors) {
      setLoading(true);
      onFinish(words, numbers);
    }
  };

  const onSubmit = (e?: React.FormEvent<HTMLFormElement>) => {
    e?.preventDefault();
    clearErrors();
    if (step === 1) {
      validateWords();
    } else {
      validateNumbers();
    }
  };

  const onWordInputChange = (value: string, index: number) => {
    setWords((w) => {
      const temp = w;
      temp[index].value = value;
      return [...temp];
    });
  };

  const onNumberInputChange = (value: string, index: number) => {
    setNumbers((n) => {
      const temp = n;
      temp[index].value = value;
      return [...temp];
    });
  };

  const addWord = () =>
    setWords((w) => [...w, { value: "", error: undefined }]);
  const removeWord = (index: number) =>
    setWords((w) => {
      const temp = w;
      temp.splice(index, 1);
      return [...temp];
    });

  const addNumber = () =>
    setNumbers((n) => [...n, { value: "", error: undefined }]);
  const removeNumber = (index: number) =>
    setNumbers((n) => {
      const temp = n;
      temp.splice(index, 1);
      return [...temp];
    });

  return {
    step,
    setStep,
    words,
    numbers,
    loading,
    nextStep,
    prevStep,
    onSubmit,
    onWordInputChange,
    onNumberInputChange,
    addWord,
    removeWord,
    addNumber,
    removeNumber,
  };
};
