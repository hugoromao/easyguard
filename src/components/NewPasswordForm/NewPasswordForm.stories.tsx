import type { Meta, StoryObj } from "@storybook/react";

import NewPasswordForm from ".";
import { SnackbarProvider } from "notistack";

const meta: Meta<typeof NewPasswordForm> = {
  component: NewPasswordForm,
  decorators: [
    (Story) => (
      <SnackbarProvider>
        <Story />
      </SnackbarProvider>
    ),
  ],
};

export default meta;

type Story = StoryObj<typeof NewPasswordForm>;

export const Primary: Story = { args: { active: true, setActive: () => ({}) } };
