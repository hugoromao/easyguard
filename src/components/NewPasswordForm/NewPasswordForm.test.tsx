import { render, screen, fireEvent } from "@/utils/test-utils";
import NewPasswordForm from ".";

describe("<NewPasswordForm onFinish={() => ({})}", () => {
  it("should not render is not active", () => {
    const { container } = render(
      <NewPasswordForm
        onFinish={() => ({})}
        isOpen={false}
        onOpenChange={() => ({})}
      />
    );
    expect(container.firstChild?.firstChild).toBeNull();
  });

  it("should render correctly", () => {
    const form = render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    expect(screen.getAllByRole("textbox").length).toBe(4);
    expect(screen.getByRole("button", { name: "next" }));

    expect(form).toMatchSnapshot();
  });

  it("should throw an error if the 'words' fields are empty.", () => {
    render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    fireEvent.click(screen.getByRole("button", { name: /next/i }));
    expect(screen.getAllByText(/Este campo é obrigatório/u).length).toEqual(4);
  });

  it("should throw an error if the 'words' have small words.", () => {
    render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    fireEvent.change(screen.getByLabelText("Palavra 1"), {
      target: { value: "ab" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 2"), {
      target: { value: "Palavra 2" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 3"), {
      target: { value: "Palavra 3" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 4"), {
      target: { value: "Palavra 4" },
    });
    fireEvent.click(screen.getByRole("button", { name: /next/i }));
    expect(screen.getByText(/A palavra “ab” é menor que três caracteres\./i));
  });

  it("should throw an error if the 'words' have common words.", () => {
    render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    fireEvent.change(screen.getByLabelText("Palavra 1"), {
      target: { value: "comum" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 2"), {
      target: { value: "Palavra 2" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 3"), {
      target: { value: "Palavra 3" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 4"), {
      target: { value: "Palavra 4" },
    });
    fireEvent.click(screen.getByRole("button", { name: /next/i }));
    expect(screen.getByText(/comum é uma palavra muito comum/i));
  });

  it("should throw an error if the 'words' have keyboard patterns.", () => {
    render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    fireEvent.change(screen.getByLabelText("Palavra 1"), {
      target: { value: "qwerty" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 2"), {
      target: { value: "Palavra 2" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 3"), {
      target: { value: "Palavra 3" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 4"), {
      target: { value: "Palavra 4" },
    });
    fireEvent.click(screen.getByRole("button", { name: /next/i }));
    expect(screen.getByText(/Não utilize padrões de teclado/i));
  });

  it('should add a textbox field when the "Adicionar palavra" button is clicked', () => {
    render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    fireEvent.click(screen.getByRole("button", { name: /Adicionar palavra/i }));

    expect(screen.getAllByRole("textbox").length).toBe(5);
  });

  it("should remove a textbox field when the trash button is clicked", () => {
    render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    const addTextboxButton = screen.getByRole("button", {
      name: /Adicionar palavra/i,
    });

    fireEvent.click(addTextboxButton);
    fireEvent.click(screen.getByRole("button", { name: "remove" }));

    expect(screen.getAllByRole("textbox").length).toBe(4);
  });

  it("should render the next step correctly", () => {
    render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    fireEvent.change(screen.getByLabelText("Palavra 1"), {
      target: { value: "Palavra 1" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 2"), {
      target: { value: "Palavra 2" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 3"), {
      target: { value: "Palavra 3" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 4"), {
      target: { value: "Palavra 4" },
    });

    fireEvent.click(screen.getByRole("button", { name: "next" }));

    expect(screen.getByText("Última etapa"));
    expect(
      screen.getByText(
        "Agora vamos precisar de dois números significativos para você. Não vale usar datas de nascimento, idade ou qualquer informação muito óbvia sobre você."
      )
    );
    expect(screen.getAllByRole("spinbutton").length).toBe(2);
    expect(screen.getByRole("button", { name: "Adicionar número" }));
    expect(screen.getByRole("button", { name: "next" }));
  });

  it('should go back to the previous step when "Go Back" button is clicked', () => {
    render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    fireEvent.change(screen.getByLabelText("Palavra 1"), {
      target: { value: "Palavra 1" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 2"), {
      target: { value: "Palavra 2" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 3"), {
      target: { value: "Palavra 3" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 4"), {
      target: { value: "Palavra 4" },
    });

    fireEvent.click(screen.getByRole("button", { name: "next" }));
    fireEvent.click(screen.getByRole("button", { name: "goBack" }));

    expect(screen.getByText("Vamos criar uma nova senha!"));
  });

  it("should throw an error if the 'numbers' fields are empty.", () => {
    render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    fireEvent.change(screen.getByLabelText("Palavra 1"), {
      target: { value: "Palavra 1" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 2"), {
      target: { value: "Palavra 2" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 3"), {
      target: { value: "Palavra 3" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 4"), {
      target: { value: "Palavra 4" },
    });

    fireEvent.click(screen.getByRole("button", { name: "next" }));
    fireEvent.click(screen.getByRole("button", { name: "next" }));

    expect(screen.getAllByText(/Este campo é obrigatório/u).length).toEqual(2);
  });

  it('should add a textbox field when the "Adicionar número" button is clicked', () => {
    render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    fireEvent.change(screen.getByLabelText("Palavra 1"), {
      target: { value: "Palavra 1" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 2"), {
      target: { value: "Palavra 2" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 3"), {
      target: { value: "Palavra 3" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 4"), {
      target: { value: "Palavra 4" },
    });

    fireEvent.click(screen.getByRole("button", { name: "next" }));
    fireEvent.click(screen.getByRole("button", { name: "Adicionar número" }));

    expect(screen.getAllByRole("spinbutton").length).toBe(3);
  });

  it("should remove a textbox field when the trash button is clicked", () => {
    render(
      <NewPasswordForm onFinish={() => ({})} isOpen onOpenChange={() => ({})} />
    );

    fireEvent.change(screen.getByLabelText("Palavra 1"), {
      target: { value: "Palavra 1" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 2"), {
      target: { value: "Palavra 2" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 3"), {
      target: { value: "Palavra 3" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 4"), {
      target: { value: "Palavra 4" },
    });

    fireEvent.click(screen.getByRole("button", { name: "next" }));
    fireEvent.click(screen.getByRole("button", { name: "Adicionar número" }));
    fireEvent.click(screen.getByRole("button", { name: "remove" }));

    expect(screen.getAllByRole("spinbutton").length).toBe(2);
  });

  it("should call onSubmitPasswordForm", () => {
    const onFinish = jest.fn();

    render(
      <NewPasswordForm onFinish={onFinish} isOpen onOpenChange={() => ({})} />
    );

    fireEvent.change(screen.getByLabelText("Palavra 1"), {
      target: { value: "Palavra 1" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 2"), {
      target: { value: "Palavra 2" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 3"), {
      target: { value: "Palavra 3" },
    });
    fireEvent.change(screen.getByLabelText("Palavra 4"), {
      target: { value: "Palavra 4" },
    });

    fireEvent.click(screen.getByRole("button", { name: "next" }));

    fireEvent.change(screen.getByLabelText("Número 1"), {
      target: { value: 123 },
    });
    fireEvent.change(screen.getByLabelText("Número 2"), {
      target: { value: 456 },
    });

    fireEvent.click(screen.getByRole("button", { name: "next" }));

    expect(onFinish).toHaveBeenCalledTimes(1);
    expect(onFinish).toHaveBeenCalledWith(
      [
        { error: undefined, value: "Palavra 1" },
        { error: undefined, value: "Palavra 2" },
        { error: undefined, value: "Palavra 3" },
        { error: undefined, value: "Palavra 4" },
      ],
      [
        { error: undefined, value: "123" },
        { error: undefined, value: "456" },
      ]
    );
  });
});
