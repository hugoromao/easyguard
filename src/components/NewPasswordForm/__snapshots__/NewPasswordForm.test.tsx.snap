// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewPasswordForm onFinish={() => ({})} should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div
      aria-hidden="true"
    >
      <div
        data-overlay-container="true"
      />
    </div>
    <span
      aria-hidden="true"
      data-focus-scope-start="true"
      hidden=""
    />
    <div
      tabindex="-1"
    >
      <div
        aria-hidden="true"
        class="z-50 backdrop-blur-md backdrop-saturate-150 bg-overlay/30 w-screen h-screen fixed inset-0"
        style="opacity: 0; will-change: opacity;"
      />
      <div
        class="flex w-screen fixed inset-0 z-50 overflow-x-auto justify-center h-[--visual-viewport-height] items-end sm:items-center [--scale-enter:100%] [--scale-exit:100%] [--slide-enter:0px] [--slide-exit:80px] sm:[--scale-enter:100%] sm:[--scale-exit:103%] sm:[--slide-enter:0px] sm:[--slide-exit:0px]"
        data-slot="wrapper"
        style="--visual-viewport-height: 768px; opacity: 0; will-change: transform,opacity; transform: translateY(var(--slide-exit)) scale(var(--scale-exit));"
      >
        <section
          aria-describedby=":r5:"
          aria-labelledby=":r4:"
          aria-modal="true"
          class="flex flex-col relative z-50 w-full box-border bg-content1 outline-none mx-1 my-1 sm:mx-6 sm:my-16 max-w-md rounded-large shadow-small overflow-y-hidden"
          data-dismissable="true"
          data-open="true"
          id=":r3:"
          role="dialog"
          tabindex="-1"
        >
          <div
            style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
          >
            <button
              aria-label="Dismiss"
              id="react-aria-:r8:"
              style="width: 1px; height: 1px;"
              tabindex="-1"
            />
          </div>
          <button
            aria-label="Close"
            class="absolute appearance-none select-none top-1 right-1 rtl:left-1 rtl:right-[unset] p-2 text-foreground-500 rounded-full hover:bg-default-100 active:bg-default-200 tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2"
            role="button"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="none"
              focusable="false"
              height="1em"
              role="presentation"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="M18 6L6 18M6 6l12 12"
              />
            </svg>
          </button>
          <header
            class="flex py-4 px-6 flex-initial text-[24px] font-semibold"
            id=":r4:"
          >
            Etapa 1
          </header>
          <div
            class="flex flex-1 flex-col gap-3 px-6 py-2 pb-0"
            id=":r5:"
          >
            <form
              autocomplete="off"
            >
              <div
                class="flex flex-col gap-2"
              >
                <p
                  class="mb-2"
                >
                  Para começar, vamos precisar de quatro palavras (ou mais). Quanto mais sem sentido, melhor! Você pode tentar criar uma história bizarra com as palavras, como: “corajoso, buriti, do mato, fez vestibular”.
                </p>
                <span
                  class="flex gap-2"
                >
                  <div
                    class="group flex flex-col data-[hidden=true]:hidden w-full"
                    data-has-elements="true"
                    data-has-label="true"
                    data-has-value="true"
                    data-slot="base"
                  >
                    <div
                      class="relative w-full inline-flex tap-highlight-transparent shadow-sm bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 min-h-8 rounded-small flex-col items-start justify-center gap-0 transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background h-12 py-1.5 px-3"
                      data-slot="input-wrapper"
                      style="cursor: text;"
                    >
                      <label
                        class="absolute z-10 pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block text-foreground-500 cursor-text will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-default-600 group-data-[filled-within=true]:pointer-events-auto group-data-[filled-within=true]:scale-85 text-small group-data-[filled-within=true]:-translate-y-[calc(50%_+_theme(fontSize.tiny)/2_-_8px)] pe-2 max-w-full text-ellipsis overflow-hidden"
                        data-slot="label"
                        for="react-aria-:r9:"
                        id="react-aria-:ra:"
                      >
                        Palavra 1
                      </label>
                      <div
                        class="inline-flex w-full items-center h-full box-border group-data-[has-label=true]:items-end pb-0.5"
                        data-slot="inner-wrapper"
                      >
                        <input
                          aria-label="Palavra 1"
                          aria-labelledby="react-aria-:r9: react-aria-:ra:"
                          class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-small group-data-[has-value=true]:text-default-foreground"
                          data-slot="input"
                          id="react-aria-:r9:"
                          name="Palavra 1"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </span>
                <span
                  class="flex gap-2"
                >
                  <div
                    class="group flex flex-col data-[hidden=true]:hidden w-full"
                    data-has-elements="true"
                    data-has-label="true"
                    data-has-value="true"
                    data-slot="base"
                  >
                    <div
                      class="relative w-full inline-flex tap-highlight-transparent shadow-sm bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 min-h-8 rounded-small flex-col items-start justify-center gap-0 transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background h-12 py-1.5 px-3"
                      data-slot="input-wrapper"
                      style="cursor: text;"
                    >
                      <label
                        class="absolute z-10 pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block text-foreground-500 cursor-text will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-default-600 group-data-[filled-within=true]:pointer-events-auto group-data-[filled-within=true]:scale-85 text-small group-data-[filled-within=true]:-translate-y-[calc(50%_+_theme(fontSize.tiny)/2_-_8px)] pe-2 max-w-full text-ellipsis overflow-hidden"
                        data-slot="label"
                        for="react-aria-:re:"
                        id="react-aria-:rf:"
                      >
                        Palavra 2
                      </label>
                      <div
                        class="inline-flex w-full items-center h-full box-border group-data-[has-label=true]:items-end pb-0.5"
                        data-slot="inner-wrapper"
                      >
                        <input
                          aria-label="Palavra 2"
                          aria-labelledby="react-aria-:re: react-aria-:rf:"
                          class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-small group-data-[has-value=true]:text-default-foreground"
                          data-slot="input"
                          id="react-aria-:re:"
                          name="Palavra 2"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </span>
                <span
                  class="flex gap-2"
                >
                  <div
                    class="group flex flex-col data-[hidden=true]:hidden w-full"
                    data-has-elements="true"
                    data-has-label="true"
                    data-has-value="true"
                    data-slot="base"
                  >
                    <div
                      class="relative w-full inline-flex tap-highlight-transparent shadow-sm bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 min-h-8 rounded-small flex-col items-start justify-center gap-0 transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background h-12 py-1.5 px-3"
                      data-slot="input-wrapper"
                      style="cursor: text;"
                    >
                      <label
                        class="absolute z-10 pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block text-foreground-500 cursor-text will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-default-600 group-data-[filled-within=true]:pointer-events-auto group-data-[filled-within=true]:scale-85 text-small group-data-[filled-within=true]:-translate-y-[calc(50%_+_theme(fontSize.tiny)/2_-_8px)] pe-2 max-w-full text-ellipsis overflow-hidden"
                        data-slot="label"
                        for="react-aria-:rj:"
                        id="react-aria-:rk:"
                      >
                        Palavra 3
                      </label>
                      <div
                        class="inline-flex w-full items-center h-full box-border group-data-[has-label=true]:items-end pb-0.5"
                        data-slot="inner-wrapper"
                      >
                        <input
                          aria-label="Palavra 3"
                          aria-labelledby="react-aria-:rj: react-aria-:rk:"
                          class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-small group-data-[has-value=true]:text-default-foreground"
                          data-slot="input"
                          id="react-aria-:rj:"
                          name="Palavra 3"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </span>
                <span
                  class="flex gap-2"
                >
                  <div
                    class="group flex flex-col data-[hidden=true]:hidden w-full"
                    data-has-elements="true"
                    data-has-label="true"
                    data-has-value="true"
                    data-slot="base"
                  >
                    <div
                      class="relative w-full inline-flex tap-highlight-transparent shadow-sm bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 min-h-8 rounded-small flex-col items-start justify-center gap-0 transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background h-12 py-1.5 px-3"
                      data-slot="input-wrapper"
                      style="cursor: text;"
                    >
                      <label
                        class="absolute z-10 pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block text-foreground-500 cursor-text will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-default-600 group-data-[filled-within=true]:pointer-events-auto group-data-[filled-within=true]:scale-85 text-small group-data-[filled-within=true]:-translate-y-[calc(50%_+_theme(fontSize.tiny)/2_-_8px)] pe-2 max-w-full text-ellipsis overflow-hidden"
                        data-slot="label"
                        for="react-aria-:ro:"
                        id="react-aria-:rp:"
                      >
                        Palavra 4
                      </label>
                      <div
                        class="inline-flex w-full items-center h-full box-border group-data-[has-label=true]:items-end pb-0.5"
                        data-slot="inner-wrapper"
                      >
                        <input
                          aria-label="Palavra 4"
                          aria-labelledby="react-aria-:ro: react-aria-:rp:"
                          class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-small group-data-[has-value=true]:text-default-foreground"
                          data-slot="input"
                          id="react-aria-:ro:"
                          name="Palavra 4"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </span>
              </div>
              <footer
                class="flex-row gap-2 py-4 justify-end flex items-center sticky bottom-0 px-0 bg-white z-10 mt-4"
              >
                <button
                  class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent data-[hover=true]:bg-default/40 w-fit text-emerald-600"
                  type="button"
                >
                  Adicionar palavra
                </button>
                <button
                  aria-label="next"
                  class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-medium gap-3 rounded-large px-0 !gap-0 transition-transform-colors-opacity motion-reduce:transition-none text-default-foreground min-w-12 w-12 h-12 data-[hover=true]:opacity-hover bg-emerald-500"
                  type="submit"
                >
                  <svg
                    aria-hidden="true"
                    class="fill-white"
                    data-slot="icon"
                    fill="currentColor"
                    height="24"
                    viewBox="0 0 20 20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M3 10a.75.75 0 0 1 .75-.75h10.638L10.23 5.29a.75.75 0 1 1 1.04-1.08l5.5 5.25a.75.75 0 0 1 0 1.08l-5.5 5.25a.75.75 0 1 1-1.04-1.08l4.158-3.96H3.75A.75.75 0 0 1 3 10Z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
              </footer>
            </form>
          </div>
          <div
            style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
          >
            <button
              aria-label="Dismiss"
              id="react-aria-:rt:"
              style="width: 1px; height: 1px;"
              tabindex="-1"
            />
          </div>
        </section>
      </div>
    </div>
    <span
      aria-hidden="true"
      data-focus-scope-end="true"
      hidden=""
    />
  </body>,
  "container": <div
    aria-hidden="true"
  >
    <div
      data-overlay-container="true"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
